contrib/epee/include/net/levin_protocol_handler_async.h

```diff
@@ -860,18 +856,19 @@ int async_protocol_handler_config<t_connection_context>::invoke_async(int comman
 template<class t_connection_context> template<class callback_t>
 bool async_protocol_handler_config<t_connection_context>::foreach_connection(const callback_t &cb)
 {
-  CRITICAL_REGION_LOCAL(m_connects_lock);
   std::vector<typename connections_map::mapped_type> conn;
-  conn.reserve(m_connects.size());
 
   auto scope_exit_handler = misc_utils::create_scope_leave_handler([&conn]{
     for (auto &aph: conn)
       aph->finish_outer_call();
   });
 
+  CRITICAL_REGION_BEGIN(m_connects_lock);
+  conn.reserve(m_connects.size());
   for (auto &e: m_connects)
     if (e.second->start_outer_call())
       conn.push_back(e.second);
+  CRITICAL_REGION_END()
```

Comment:
I have a comment about this in my conversion to `shared_ptr`/`weak_ptr` patch - I would revert this line right here. Several of the callbacks appear to assume that they have some exclusive access with respect to other calls to the `foreach_connection`. So unless you've stepped through the logic, its safer to revert this. Which might be an issue for deadock, which means recursive mutex.

Reply:
This is long, but please read the first paragraph to at least understand my point -

There's two different potential deadlock cases in this function, and these tests only cover one of them (presumably because @selsta stacktraces only contained this case). If you move this lock release to the bottom, the tests still pass. The "primary" deadlock case is the connection releases occurring before the lock release (the order of which was changed above). This occurs when `foreach_connection` has the last reference to a connection.

The second deadlock case is dependent on callback behavior; the callback tries to acquire `m_connects_lock` somehow. Unfortunately, there are "naked" reads/writes occurring in these callbacks (see below), so its not obvious that this is an improvement (it might be, but I think this is subjective). My recommendation was/is to not release the lock here, and to keep the change above where the scoped release order changes, since (afaik) there are no known deadlock cases but known race conditions. Also, there's no test for this right now.

-----
"Proof":

There's already naked read/write calls in `foreach_connection` - here's an [instance of `m_state` being written](https://github.com/monero-project/monero/blob/master/src/cryptonote_protocol/cryptonote_protocol_handler.inl#L1690), with many [read calls inside of `for_each_connection`](https://github.com/monero-project/monero/blob/master/src/cryptonote_protocol/cryptonote_protocol_handler.inl#L1771) (which forwards to `foreach_connection`) within this file. There are more read calls that just this one instance.

This particular variable is already an issue elsewhere, but removing the lock hold is exacerbating the problem. Unless the proof you want is showing that this is worse than deadlock - which is difficult because this type of UB can be harder to identify than deadlock. Thread-saniziter already complains constantly about the context object - the code was mosty written with the assumption that the context was exclusively being accessed.

Reply:
```
The above thread unsafe access to connection context must be fixed separately
  since it existed before this patch and will continue to exist after.
Provided unit test has been updated to catch both ways to encounter deadlock
  within async_protocol_handler_config<T>::foreach_connection().
```

Reply:
> The above thread unsafe access to connection context must be fixed separately
  since it existed before this patch and will continue to exist after.

I agree this is where the code needs to eventually be, but as stated above, whether this better now is subjective since it adds more race conditions but fixes no known deadlocks caused by real callbacks. Deadlocks stop progress and can be backtraced; race conditions are harder to diagnose.

If others seem fine with this, then this is difficult to object. My recommendation would be to merge on `master` first for some testnet stressing before merging onto 0.17.

> Provided unit test has been updated to catch both ways to encounter deadlock
  within async_protocol_handler_config<T>::foreach_connection().

Might as well add one to `for_connection`, the situation is identical.

Reply:
```
Provided unit test has been updated to catch both ways to encounter deadlock
  within async_protocol_handler_config<T>::for_connection().
```